Explore: Allow switching between metrics and logs

marefr opened the pull request:
**What this PR does / why we need it**:
Adds basic support for switching between Metrics and Logs. Currently only test datasource that supports both Metrics and Logs.

- Basic support for switching between Metrics and Logs
 - Introduces `mode` and `supportedModes` properties on `ExploreItemState`
 - `supportedModes` will include Metrics if datasource has `"metrics": true` in plugin.json
 - `supportedModes` will include Logs if datasource has `"logs": true` in plugin.json
 - Metrics and Logs switcher currently only visible if `supportedModes` includes more than one item
 - If loading explore for a datasource that supports both Metrics and Logs then Metrics are selected as the default.
 - Enables logs for test datasource
- Initial UX for switching between Metrics and Logs according to mockups (see issue)

Open questions:
 - For the datasources supporting both Metrics and Logs do we need to introduce some new setting in plugin.json to define which one should be the default or should logs always be the preferred in Explore?
   - After discussion Metrics should be the default

![image](https://user-images.githubusercontent.com/1668778/57637953-011d4c00-75ad-11e9-8f9b-473e7c110c4d.png)

![image](https://user-images.githubusercontent.com/1668778/57637960-05e20000-75ad-11e9-913b-402d748df4b3.png)

![image](https://user-images.githubusercontent.com/1668778/57637964-09758700-75ad-11e9-8747-33980bd6f069.png)

![image](https://user-images.githubusercontent.com/1668778/57637977-0c707780-75ad-11e9-806c-18a69e4dd591.png)

![Peek 2019-05-13 18-26](https://user-images.githubusercontent.com/1668778/57638016-20b47480-75ad-11e9-9a3e-7e199c1d3420.gif)

**Which issue(s) this PR fixes**:
Closes #16808

**Special notes for your reviewer**:

marefr commented on public/app/features/panel/query_ctrl.ts:
```diff
@@ -8,10 +9,15 @@ export class QueryCtrl {
     this.isLastQuery = _.indexOf(this.panel.targets, this.target) === this.panel.targets.length - 1;
+
+    if (!this.queryType) {
+      this.queryType = QueryType.Metrics;
```
Default to Metrics for dashboards for now.

torkelo commented:
There is something with the term queryType that feels awkward as it's more a visualization/ui option / toggle. But not sure what better alternatives there are :)

marefr commented:
Yeah I've been struggling with the naming as well :smile: But since this change will allow us to in theory load different query editors based on query type I thought that was at least better than dataType that I initially tried. ResultType is another idea which denotes what you want to show. DataSourceType or SourceType is another idea which denotes what's the source type of the data you want to show.

Suggestions appreciated

torkelo commented:
The problem with all of those is that is the mode would (at least in some cases) not change the query or the response type. Just how it's visualized. It could change the query UI and most cases the query model I guess. 

Some other ideas (but not sure I prefer these over QueryType)

* QueryMode
* DataVizMode
* DataMode

steven-sheehy commented:
Would it also make sense for the Logs to be split into separate collapsible Graph and Table panels as the metrics is to be more consistent?

hugohaggmark commented:
Great work, looks very good so far. 👍 

I'm also not happy with QueryType because that suggests that should be a prop on `DataQuery` not on the Editor. Maybe is should be more like `VisualizationMode` and maybe this should only be part of `ExploreQueryFieldProps` and not part of the `QueryEditorProps`?

Or are you thinking we're going to introduce this on Panels as well?

hugohaggmark commented:
Found a strange behaviour:

1. Goto Explore
2. Choose TestData, `Random Walk` is automatically chosen
3. Automatically switches to `Logs`, and no Log results show 👍 
4. Change to `Random Walk Table`
5. Shows Graph and Logs result graph 👍 
![image](https://user-images.githubusercontent.com/562238/57507572-7d8b0300-72ff-11e9-8e8a-f2297d6d3ecf.png)

6. Add new QueryRow, `Random Walk` is automatically chosen
7. Graph has now more data then before, strange
![image](https://user-images.githubusercontent.com/562238/57507610-93002d00-72ff-11e9-9f37-c85e3a88fb0c.png)

8. Remove the ´QueryRow´ that contains `Random Walk Table`
9. Logs result is now empty again, strange because in step 6 more data was added to graph suggesting there was data
![image](https://user-images.githubusercontent.com/562238/57507653-aad7b100-72ff-11e9-86b2-bbcfcb528dfa.png)

marefr committed:
wip

marefr commented:
> Found a strange behaviour:

@hugohaggmark yeah that "strange behavior" is intentional :smile:  Some datasources, for example Elasticsearch, can return both log lines and histogram/metrics in the same request where Loki currently only return log lines and a histogram is created based on the log lines. Based on how this is implemented right now the histogram will only be shown in graph if there are log lines returned. I personally think that it's weird to be able to show graph without logs since the graph are showing a histogram of log lines over time.

marefr commented:
@torkelo @hugohaggmark Thanks for the feedback!

How about `ExploreKind` or `ExploreType` and limit this concept to Explore to start with?

marefr commented:
@steven-sheehy 

> Would it also make sense for the Logs to be split into separate collapsible Graph and Table panels as the metrics is to be more consistent?

I would say no since graph and table fulfills two things looking at the same data, details and distribution over time. Are you referring to being able to collapse graph when looking at logs or what would be the use case you're referring to?

steven-sheehy commented:
@marefr Yes, to be able to collapse the graph when looking at logs. Right now the graph is pretty useless for Loki since it doesn't return a histogram and the graph is calculated client side. I have some ideas on how to provide that histogram in Loki but waiting on the log label parsing support to stabilize first.

> I would say no since graph and table fulfills two things looking at the same data, details and distribution over time.

One can argue you that for Promethueus metrics in Explore the graph and table fulfills two things looking at the same data, details and distribution over time. And each one is collapsible.

torkelo commented:
> How about ExploreKind or ExploreType and limit this concept to Explore to start with?

Maybe ExploreMode?  since the main thing that changes will be the query UI & visualization.

marefr committed:
change naming from queryType to mode

Moves mode (Metric, Logs) selection to the left of datasource picker and add
some quick styling.
Instead of supportedQueryTypes and queryType in ExploreItemState we're trying
supportedModes and mode instead with the type ExploreMode.

marefr commented:
Pushed some updates:
- Renamed properties `supportedQueryTypes` and `queryType` to `supportedModes` and `mode` on `ExploreItemState`.
- Renamed action type `changeQueryTypeAction` to `changeModeAction`.

Felt really weird to name the property `exploreMode` on `ExploreItemState` since we're in the context of Explore. 

Please note that these changes are now limited to the scope of Explore to reduce the scope of this change and to move this forward.

marefr commented:
Should add that I still find the `mode` property a bit weird. Was on my way to use `displayMode`, but didn't after discussions with Torkel :smile:

marefr commented:
@torkelo any input on my latest changes?

torkelo commented:
Think they look ok

marefr commented:
Ok is okay with me :smile: Thanks

marefr committed:
only trigger change in ToggleButton if not selected

marefr committed:
set correct mode if datasource only supports logs

hugohaggmark reviewed (approved):
Great work, LGTM!

One minor thing, haven't looked when this was introduced but responsive view for ipad/iphones look like there is an unused area on the left. Maybe part of another PR to make use of this empty space as well.

![image](https://user-images.githubusercontent.com/562238/57826906-95520380-77a5-11e9-89ba-baf3860bcfa3.png)

marefr commented:
Thanks. Same problem with empty space on master so suggest fixing that in a separate PR as you suggest.
